### PR TITLE
Add support for reloadable FileResourceGroupConfigurationManager

### PR DESCRIPTION
**Pull request by nezihyigitbasi**
This PR adds:
- a read only rest endpoint for resource groups
- reload support for `FileResourceGroupConfigurationManager`

### CODE REVIEW

**Comment by nezihyigitbasi** [comment]
@cberner can you please take a look?

File: presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/FileResourceGroupConfigurationManager.java [hunk 1]
```diff
@@ -81,6 +117,62 @@ public void configure(ResourceGroup group, SelectionContext context)
...
+                            listeners.stream().forEach(listener -> listener.accept(null));
+                        }
+                    }
+                }
+                catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    Throwables.propagate(e);
+                }
+                catch (Throwable t) {
+                    log.warn("Exception while reloading changes: %s", t.getMessage());
```
Comment (electrum) [inline presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/FileResourceGroupConfigurationManager.java]: Log the exception so we get the full stack trace:
```java
log.warn(e, "Reloading resource group file failed");

File: presto-main/src/main/java/com/facebook/presto/server/ResourceManagerResource.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    public ResourceManagerResource(ResourceGroupManager resourceGroupManager)
+    {
+        this.resourceGroupManager = requireNonNull(resourceGroupManager, "resourceGroupManager is null");
+    }
+
+    @GET
+    @Produces(APPLICATION_JSON)
+    public List<ResourceGroupInfo> getGroups()
+    {
+        return resourceGroupManager
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/ResourceManagerResource.java]: It's more canonical to write this as
```java
return resourceGroupManager.getRootResourceGroups().stream()
        .map(ResourceGroup::getInfo)
        .collect(toImmutableList());
```
Put `.stream()` on the first line and use method reference

File: presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/FileResourceGroupConfigurationManager.java [hunk 3]
```diff
@@ -81,6 +117,62 @@ public void configure(ResourceGroup group, SelectionContext context)
...
+                }
+                finally {
+                    if (key != null) {
+                        if (!key.reset()) {
```
Comment (electrum) [inline presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/FileResourceGroupConfigurationManager.java]: Why would this occur? Is it a bug?

**Comment by nezihyigitbasi** [comment]
addressed @electrum's comments

File: presto-main/src/main/java/com/facebook/presto/server/ResourceManagerResource.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import javax.ws.rs.Produces;
+
+import java.util.List;
+
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static java.util.Objects.requireNonNull;
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
+
+@Path("/v1/resource-manager")
+public class ResourceManagerResource
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/server/ResourceManagerResource.java]: Could you split this into two commits? One that adds this Resource (and whatever methods it needs), and one that adds the reloading capabilities. We like to put distinct features in separate commits

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroupManager.java [hunk 5]
```diff
@@ -156,6 +181,19 @@ public void setConfigurationManager(String name, Map<String, String> properties)
         ResourceGroupConfigurationManager configurationManager = configurationManagerFactory.create(ImmutableMap.copyOf(properties), () -> memoryPoolManager);
         checkState(this.configurationManager.compareAndSet(null, configurationManager), "configurationManager already set");
 
+        if (configurationManager instanceof ReloadingConfigurationManager) {
+            ((ReloadingConfigurationManager) configurationManager).addConfigReloadListener((ignored) -> {
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroupManager.java]: I don't think this listener approach is going to work. By clearing the `groups` and `rootGroups` previous queries that were submitted to a resource group, "a", will be running in that old object, but new ones will go to a new copy of "a", this could lead to the group going over quota.

I'm pretty sure you don't need to make any changes to this class actually. I designed the `InternalResourceGroup` objects to be thread-safe and re-configurable, so the `FileResourceGroupConfigurationManager` can just retain a reference to them (see the docs for `ResourceGroupConfigurationManager` which explicitly allow this) and then reconfigure the groups whenever a change to the file is detected

**Comment by nezihyigitbasi** [comment]
@cberner I pushed a commit for adding a REST endpoint.  I am still working on the reloading part (working on some deadlock issue that I mentioned in #6755). Once I have something concrete I will let you know.

File: presto-main/src/main/java/com/facebook/presto/server/ResourceManagerResource.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+
+import java.util.List;
+
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static java.util.Objects.requireNonNull;
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
+
+@Path("/v1/resource-manager")
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/server/ResourceManagerResource.java]: I'd call this `/v1/resource-group`. @dain could you take a look at this file and the JSON format of `ResourceGroupInfo` as well, since this will be a external API?

File: presto-resource-group-managers/src/test/java/com/facebook/presto/resourceGroups/TestingResourceGroup.java [hunk 7]
```diff
@@ -153,4 +154,10 @@ public void setJmxExport(boolean export)
     {
         jmxExport = export;
     }
+
+    @Override
+    public ResourceGroupInfo getInfo()
+    {
+        return null;
```
Comment (cberner) [inline presto-resource-group-managers/src/test/java/com/facebook/presto/resourceGroups/TestingResourceGroup.java]: Probably better to throw an `UnsupportedOperationException` here

**Comment by cberner** [comment]
Can you add a test similar to `TestExecuteResource` for this new resource?

**Comment by cberner** [comment]
Couple more comments

**Comment by nezihyigitbasi** [comment]
@cberner updated. Depends on #6796.

**Comment by nezihyigitbasi** [comment]
@cberner rebased.

**Comment by nezihyigitbasi** [comment]
rebased

**Comment by nezihyigitbasi** [comment]
@cberner rebased & added a simple test for the new REST endpoint (`TestResourceManagerResource`).

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/ResourceGroupManager.java [hunk 8]
```diff
@@ -25,4 +29,6 @@
     void addConfigurationManagerFactory(ResourceGroupConfigurationManagerFactory factory);
 
     void loadConfigurationManager() throws Exception;
+
+    List<ResourceGroup> getRootResourceGroups();
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/ResourceGroupManager.java]: I find this confusing. Currently ResourceGroup interface doesn't suggest that there has to be a root resource group. The interface itself doesn't have any indication about the structure. So it's confusing to have ResourceGroupManager interface to have an API for getRootResourceGroups (which indicate that resource groups are arranged in a tree structure). If we are adding getRootResourceGroups here, we should probably also add a getRoot in ResourceGroup interface. Especially since ResourceGroupInfo is a tree structure. Another way is to keep "root" only a concept with InternalResourceGroup, then we should name the current ResourceGroupInfo InternalResourceGroupInfo.

Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/ResourceGroupManager.java]: Although the `ResourceGroup ` interface doesn't have the notion of a root, the `ResourceGroupId` class (that you can get with `ResourceGroup::getId()`) has a `getParent()` method, which indicates there's a hierarchy/structure to the resource groups -- `InternalResourceGroupManager` uses `getParent()` to see whether a group is root or not.

Another option is to introduce `ResourceGroupManager::getResourceGroups()` instead of `ResourceGroupManager::getRootResourceGroups()` that return all the groups, and we let the caller figure out the hierarchy.
